MudTreeviewItem: Fix checkbox in multi-selection mode

henon opened the pull request:
Resolves #8940



## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

henon committed:
MudTreeViewItem: fix #8750

henon committed:
Docs: MudTreeView custom checkbox example fix #8750

Yomodo commented:
@henon I found one more issue, clicking the first node skips the nested subitems node:

![image](https://github.com/MudBlazor/MudBlazor/assets/10358198/1109462b-adb6-40ad-b00c-b72769af488d)

ScarletKuro commented:
@henon do we reopen this issue or shall we create new one?

henon commented:
No need, I'll directly PR a fix

henon commented:
> Multi selection: Selecting all now shows indeterminate state:
> 
> ![image](https://private-user-images.githubusercontent.com/10358198/329814698-85346dec-5dc6-4477-87f8-adc1d10c99e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTY1MTUsIm5iZiI6MTcxNTU5NjIxNSwicGF0aCI6Ii8xMDM1ODE5OC8zMjk4MTQ2OTgtODUzNDZkZWMtNWRjNi00NDc3LTg3ZjgtYWRjMWQxMGM5OWU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEzVDEwMzAxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0ZDdhYjE0ZDRkYmJiODFjZDQyZjA3YjE0NDVhYjBmYjA0ZDA5YThiOTgzZjMxMjY0NmQxZDI0ZTc1MGQ2ODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.w-y6qRUX3i9jZShPqlL_MDbNDADIo-w6wdMlwAuG8jw)

This is intended behavior, you shouldn't confuse this example with the custom logic one. Here selecting the parent just selects it and not its children. Conversely selecting all its children doesn't automatically select the parent. That is exactly why we have the custom checkbox example to show how to implement this feature if required

henon commented:
OK, sorry @Yomodo nevermind what I said above, selection should select children and vice versa. There is a bug in there, I can reproduce it. Whether it happens, depends on whether or not the parent was selected.

henon commented:
OK, flip-flopping back to the original statement, it isn't a bug, it is just very confusing that's all. The thing causing the confusion in this example is the fact that a directory has its own selected state which does not depend on the selection of its children. So when the parent is checked off, and you check all children, then the parent will remain unchecked for TriState=false 
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/666c01e5-301c-483e-aa44-f993025d7e32)
and indeterminate for TriState=true
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/db4a71a0-ab65-424c-8700-731f091773ed)

When the parent is checked and children are not all checked: TriState=false
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/e81f693e-59f4-4a9d-8508-bffc7ac12ec0)
When the parent is checked and children are not all checked: TriState=true
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/75a0d561-a040-4c35-bc89-5df24e7a9230)

This behavior is intended but I realize it isn't very consistent. Checking the parent automatically checks all children but unchecking a child doesn't update the parent state. I will change it. Then I guess we don't need the custom logic example any longer as the treeview will behave like it automatically.

henon commented:
I implemented the expected change in behavior so that checking/unchecking children updates the parent. 

![treeview hierarchic selection](https://github.com/MudBlazor/MudBlazor/assets/44090/3200b5cd-a41b-47ea-9bf0-0a66438bba30)

And I removed the custom checkbox example as it makes no sense any longer.

henon commented:
Just note that modifying the selection via SelectedValues does not trigger any hierarchic selection like when you click the checkboxes. So for instance, if you select only `bundle.zip` in the chipset and leave everything else off, it will not cause all subtrees to become selected. I hope that is ok.

henon commented:
@Yomodo: see PR #8957

Yomodo commented:
Yes, with correct use of the `TriState` property, it's now exactly as (I) expected. Thanks so much!
